Flaky test: `ArmeriaCentralDogmaBuilderTest.buildingWithProfile()`

Travis-CI [build#869](https://travis-ci.com/line/centraldogma/builds/119732002#L939)  failed intermittently

```java
com.linecorp.centraldogma.client.armeria.ArmeriaCentralDogmaBuilderTest > buildingWithProfile FAILED
    org.awaitility.core.ConditionTimeoutException: Assertion condition defined as a lambda expression 
      in com.linecorp.centraldogma.client.armeria.ArmeriaCentralDogmaBuilderTest that uses com.linecorp.armeria.client.endpoint.EndpointGroup 
    Expecting:
      <[Endpoint{1.2.3.4.xip.io:36462, ipAddr=1.2.3.4, weight=1000}]>
    to contain exactly in any order:
      <[Endpoint{1.2.3.4.xip.io:36462, ipAddr=1.2.3.4, weight=1000},
        Endpoint{5.6.7.8.xip.io:8080, ipAddr=5.6.7.8, weight=1000}]>
    but could not find the following elements:
      <[Endpoint{5.6.7.8.xip.io:8080, ipAddr=5.6.7.8, weight=1000}]>
     within 10 seconds.
        at org.awaitility.core.ConditionAwaiter.await(ConditionAwaiter.java:145)
        at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:122)
        at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:32)
        at org.awaitility.core.ConditionFactory.until(ConditionFactory.java:902)
        at org.awaitility.core.ConditionFactory.untilAsserted(ConditionFactory.java:723)
        at com.linecorp.centraldogma.client.armeria.ArmeriaCentralDogmaBuilderTest.buildingWithProfile(ArmeriaCentralDogmaBuilderTest.java:69)
        Caused by:
        java.lang.AssertionError: 
        Expecting:
          <[Endpoint{1.2.3.4.xip.io:36462, ipAddr=1.2.3.4, weight=1000}]>
        to contain exactly in any order:
          <[Endpoint{1.2.3.4.xip.io:36462, ipAddr=1.2.3.4, weight=1000},
            Endpoint{5.6.7.8.xip.io:8080, ipAddr=5.6.7.8, weight=1000}]>
        but could not find the following elements:
          <[Endpoint{5.6.7.8.xip.io:8080, ipAddr=5.6.7.8, weight=1000}]>
            at com.linecorp.centraldogma.client.armeria.ArmeriaCentralDogmaBuilderTest.lambda$buildingWithProfile$0(ArmeriaCentralDogmaBuilderTest.java:72)
```

--- a/client/java-armeria/src/test/java/com/linecorp/centraldogma/client/armeria/ArmeriaCentralDogmaBuilderTest.java
+++ b/client/java-armeria/src/test/java/com/linecorp/centraldogma/client/armeria/ArmeriaCentralDogmaBuilderTest.java
@@ -16,6 +16,9 @@
 
 package com.linecorp.centraldogma.client.armeria;
 
+import static com.linecorp.centraldogma.client.armeria.TestDnsServer.newAddressRecord;
+import static io.netty.handler.codec.dns.DnsRecordType.A;
+import static io.netty.handler.codec.dns.DnsSection.ANSWER;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.awaitility.Awaitility.await;
 import static org.mockito.ArgumentMatchers.any;
@@ -27,7 +30,11 @@ import static org.mockito.Mockito.verify;
 import java.util.List;
 import java.util.function.Consumer;
 
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
+
+import com.google.common.collect.ImmutableMap;
 
 import com.linecorp.armeria.client.ClientBuilder;
 import com.linecorp.armeria.client.ClientFactory;
@@ -36,18 +43,40 @@ import com.linecorp.armeria.client.HttpClient;
 import com.linecorp.armeria.client.endpoint.EndpointGroup;
 import com.linecorp.armeria.client.endpoint.dns.DnsAddressEndpointGroup;
 import com.linecorp.armeria.client.endpoint.healthcheck.HealthCheckedEndpointGroup;
-import com.linecorp.centraldogma.testing.internal.FlakyTest;
 
-@FlakyTest
+import io.netty.handler.codec.dns.DefaultDnsQuestion;
+import io.netty.handler.codec.dns.DefaultDnsResponse;
+
 class ArmeriaCentralDogmaBuilderTest {
 
-    // Note: This test case relies on https://sslip.io/
+    @SuppressWarnings("NotNullFieldNotInitialized")
+    private static TestDnsServer dnsServer;
+
+    @BeforeAll
+    static void setup() {
+        dnsServer = new TestDnsServer(
+                ImmutableMap.of(new DefaultDnsQuestion("1.2.3.4.foo.com.", A),
+                                new DefaultDnsResponse(0).addRecord(
+                                        ANSWER, newAddressRecord("1.2.3.4.foo.com.", "1.2.3.4")),
+                                new DefaultDnsQuestion("5.6.7.8.foo.com.", A),
+                                new DefaultDnsResponse(0).addRecord(
+                                        ANSWER, newAddressRecord("5.6.7.8.foo.com.", "5.6.7.8"))));
+    }
+
+    @AfterAll
+    static void tearDown() {
+        dnsServer.close();
+    }
 
     @Test
     void buildingWithProfile() throws Exception {
         final ArmeriaCentralDogmaBuilder b = new ArmeriaCentralDogmaBuilder();
+        b.dnsAddressEndpointGroupConfigurator(
+                configurator -> configurator.serverAddressStreamProvider(dnsServer.dnsServerList()));
         b.healthCheckIntervalMillis(0);
-        b.profile("sslip");
+        // See src/test/resources/centraldogma-client-test.json
+        b.profile("foo");
+
         try (EndpointGroup group = b.endpointGroup()) {
             assertThat(group).isNotNull();
             assertThat(group).isInstanceOf(CompositeEndpointGroup.class);
@@ -61,8 +90,8 @@ class ArmeriaCentralDogmaBuilderTest {
                 final List<Endpoint> endpoints = group.endpoints();
                 assertThat(endpoints).isNotNull();
                 assertThat(endpoints).containsExactlyInAnyOrder(
-                        Endpoint.of("1.2.3.4.sslip.io", 36462).withIpAddr("1.2.3.4"),
-                        Endpoint.of("5.6.7.8.sslip.io", 8080).withIpAddr("5.6.7.8"));
+                        Endpoint.of("1.2.3.4.foo.com", 36462).withIpAddr("1.2.3.4"),
+                        Endpoint.of("5.6.7.8.foo.com", 8080).withIpAddr("5.6.7.8"));
             });
         }
     }
@@ -87,22 +116,27 @@ class ArmeriaCentralDogmaBuilderTest {
     @Test
     void buildingWithSingleUnresolvedHost() throws Exception {
         final ArmeriaCentralDogmaBuilder b = new ArmeriaCentralDogmaBuilder();
+        b.dnsAddressEndpointGroupConfigurator(
+                configurator -> configurator.serverAddressStreamProvider(dnsServer.dnsServerList()));
         b.healthCheckIntervalMillis(0);
-        b.host("1.2.3.4.sslip.io");
+        b.host("1.2.3.4.foo.com");
+
         try (EndpointGroup endpointGroup = b.endpointGroup()) {
             endpointGroup.whenReady().join();
             assertThat(endpointGroup).isInstanceOf(DnsAddressEndpointGroup.class);
             assertThat(endpointGroup.endpoints().size()).isEqualTo(1);
-            assertThat(endpointGroup.endpoints().get(0).host()).isEqualTo("1.2.3.4.sslip.io");
+            assertThat(endpointGroup.endpoints().get(0).host()).isEqualTo("1.2.3.4.foo.com");
         }
     }
 
     @Test
     void buildingWithMultipleHosts() throws Exception {
         final ArmeriaCentralDogmaBuilder b = new ArmeriaCentralDogmaBuilder();
+        b.dnsAddressEndpointGroupConfigurator(
+                configurator -> configurator.serverAddressStreamProvider(dnsServer.dnsServerList()));
         b.healthCheckIntervalMillis(0);
-        b.host("1.2.3.4.sslip.io", 1); // Unresolved host
-        b.host("5.6.7.8.sslip.io", 2); // Another unresolved host
+        b.host("1.2.3.4.foo.com", 1); // Unresolved host
+        b.host("5.6.7.8.foo.com", 2); // Another unresolved host
         b.host("4.3.2.1", 3); // Resolved host
         b.host("8.7.6.5", 4); // Another resolved host
 
@@ -120,8 +154,8 @@ class ArmeriaCentralDogmaBuilderTest {
                 final List<Endpoint> endpoints = endpointGroup.endpoints();
                 assertThat(endpoints).isNotNull();
                 assertThat(endpoints).containsExactlyInAnyOrder(
-                        Endpoint.of("1.2.3.4.sslip.io", 1).withIpAddr("1.2.3.4"),
-                        Endpoint.of("5.6.7.8.sslip.io", 2).withIpAddr("5.6.7.8"),
+                        Endpoint.of("1.2.3.4.foo.com", 1).withIpAddr("1.2.3.4"),
+                        Endpoint.of("5.6.7.8.foo.com", 2).withIpAddr("5.6.7.8"),
                         Endpoint.of("4.3.2.1", 3),
                         Endpoint.of("8.7.6.5", 4));
             });

--- a/client/java-armeria/src/test/java/com/linecorp/centraldogma/client/armeria/TestDnsServer.java
+++ b/client/java-armeria/src/test/java/com/linecorp/centraldogma/client/armeria/TestDnsServer.java
@@ -1,0 +1,157 @@
+/*
+ * Copyright 2025 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.centraldogma.client.armeria;
+
+import static io.netty.handler.codec.dns.DnsRecordType.A;
+import static io.netty.handler.codec.dns.DnsRecordType.AAAA;
+
+import java.net.InetSocketAddress;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import com.linecorp.armeria.common.CommonPools;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.util.TransportType;
+
+import io.netty.bootstrap.Bootstrap;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInitializer;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.SimpleChannelInboundHandler;
+import io.netty.handler.codec.dns.DatagramDnsQuery;
+import io.netty.handler.codec.dns.DatagramDnsQueryDecoder;
+import io.netty.handler.codec.dns.DatagramDnsResponse;
+import io.netty.handler.codec.dns.DatagramDnsResponseEncoder;
+import io.netty.handler.codec.dns.DefaultDnsRawRecord;
+import io.netty.handler.codec.dns.DnsQuestion;
+import io.netty.handler.codec.dns.DnsRecord;
+import io.netty.handler.codec.dns.DnsResponse;
+import io.netty.handler.codec.dns.DnsResponseCode;
+import io.netty.handler.codec.dns.DnsSection;
+import io.netty.resolver.dns.DnsServerAddressStreamProvider;
+import io.netty.resolver.dns.DnsServerAddresses;
+import io.netty.util.NetUtil;
+import io.netty.util.ReferenceCountUtil;
+
+// TODO(minwoox): Make TestDnsServer in Armeria as an extension and export it.
+final class TestDnsServer implements AutoCloseable {
+
+    public static DnsRecord newAddressRecord(String name, String ipAddr) {
+        return newAddressRecord(name, ipAddr, 60);
+    }
+
+    public static DnsRecord newAddressRecord(String name, String ipAddr, long ttl) {
+        return new DefaultDnsRawRecord(
+                name, NetUtil.isValidIpV4Address(ipAddr) ? A : AAAA,
+                ttl, Unpooled.unreleasableBuffer(Unpooled.wrappedBuffer(
+                NetUtil.createByteArrayFromIpAddressString(ipAddr))));
+    }
+
+    private final Channel channel;
+    private volatile Map<DnsQuestion, DnsResponse> responses;
+
+    TestDnsServer(Map<DnsQuestion, DnsResponse> responses) {
+        this(responses, null);
+    }
+
+    TestDnsServer(Map<DnsQuestion, DnsResponse> responses,
+                  @Nullable ChannelHandler beforeDnsServerHandler) {
+        this.responses = ImmutableMap.copyOf(responses);
+
+        final Bootstrap b = new Bootstrap();
+        b.channel(TransportType.datagramChannelType(CommonPools.workerGroup()));
+        b.group(CommonPools.workerGroup());
+        b.handler(new ChannelInitializer<Channel>() {
+            @Override
+            protected void initChannel(Channel ch) throws Exception {
+                final ChannelPipeline p = ch.pipeline();
+                p.addLast(new DatagramDnsQueryDecoder());
+                p.addLast(new DatagramDnsResponseEncoder());
+                if (beforeDnsServerHandler != null) {
+                    p.addLast(beforeDnsServerHandler);
+                }
+                p.addLast(new DnsServerHandler());
+            }
+        });
+
+        channel = b.bind(NetUtil.LOCALHOST, 0).syncUninterruptibly().channel();
+    }
+
+    DnsServerAddressStreamProvider dnsServerList() {
+        final InetSocketAddress dnsServerAddr = (InetSocketAddress) channel.localAddress();
+        return hostname -> DnsServerAddresses.sequential(ImmutableList.of(dnsServerAddr)).stream();
+    }
+
+    void setResponses(Map<DnsQuestion, DnsResponse> responses) {
+        this.responses.values().forEach(DnsResponse::release);
+        this.responses = ImmutableMap.copyOf(responses);
+    }
+
+    @Override
+    public void close() {
+        if (!channel.isOpen()) {
+            return;
+        }
+
+        channel.close().syncUninterruptibly();
+        responses.values().forEach(DnsResponse::release);
+    }
+
+    private class DnsServerHandler extends SimpleChannelInboundHandler<DatagramDnsQuery> {
+        @Override
+        protected void channelRead0(ChannelHandlerContext ctx, DatagramDnsQuery query) {
+            final DnsQuestion question = query.recordAt(DnsSection.QUESTION, 0);
+            boolean responded = false;
+            for (Entry<DnsQuestion, DnsResponse> e : responses.entrySet()) {
+                final DnsQuestion q = e.getKey();
+                final DnsResponse r = e.getValue();
+                if (question.dnsClass() == q.dnsClass() &&
+                    question.name().equals(q.name()) &&
+                    question.type() == q.type()) {
+
+                    final DnsResponse res = new DatagramDnsResponse(
+                            null, query.sender(), query.id(), query.opCode(), r.code());
+
+                    if (r.count(DnsSection.QUESTION) == 0) {
+                        res.addRecord(DnsSection.QUESTION, question);
+                    }
+                    for (DnsSection section : DnsSection.values()) {
+                        final int count = r.count(section);
+                        for (int i = 0; i < count; i++) {
+                            res.addRecord(section, ReferenceCountUtil.retain(r.recordAt(section, i)));
+                        }
+                    }
+
+                    ctx.writeAndFlush(res);
+                    responded = true;
+                }
+            }
+
+            if (!responded) {
+                final DnsResponse res = new DatagramDnsResponse(
+                        null, query.sender(), query.id(), query.opCode(), DnsResponseCode.NXDOMAIN);
+                res.addRecord(DnsSection.QUESTION, question);
+                ctx.writeAndFlush(res);
+            }
+        }
+    }
+}

--- a/client/java-armeria/src/test/resources/centraldogma-profiles-test.json
+++ b/client/java-armeria/src/test/resources/centraldogma-profiles-test.json
@@ -1,14 +1,14 @@
 [
   {
-    "name": "sslip",
+    "name": "foo",
     "hosts": [
       {
-        "host": "1.2.3.4.sslip.io",
+        "host": "1.2.3.4.foo.com",
         "protocol": "http",
         "port": 36462
       },
       {
-        "host": "5.6.7.8.sslip.io",
+        "host": "5.6.7.8.foo.com",
         "protocol": "http",
         "port": 8080
       }

